File: app/src/main/java/com/fitaleks/heroesthesaurus/network/NetworkHelper.java
package com.fitaleks.heroesthesaurus.network;

import android.util.Log;

import com.fitaleks.heroesthesaurus.BuildConfig;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.TypeAdapter;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by alexander on 21.08.15.
 */
public class NetworkHelper {
    private static final String LOG_TAG = NetworkHelper.class.getSimpleName();
    private static final String MARVEL_ENDPOINT = "http://gateway.marvel.com/v1/public/";
    private static final String TAG = NetworkHelper.class.getSimpleName();

    public static MarvelFetchService getRestAdapter() {

        final Gson gson = new GsonBuilder()
                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                .registerTypeAdapterFactory(new MarvelTypeAdapterFactory())
                .create();

        final Interceptor requestInterceptor =
                new Interceptor() {
                    @Override
                    public Response intercept(Chain chain) throws IOException {
                        Request request = chain.request();
                        final HttpUrl httpUrl = request.url().newBuilder()
                                .addQueryParameter("apikey", BuildConfig.MARVEL_PUB_API_KEY)
                                .build();
                        request = request.newBuilder().url(httpUrl).build();
                        return chain.proceed(request);
                    }
                };

        final HttpLoggingInterceptor loggingInterceptor1 = new HttpLoggingInterceptor();
        loggingInterceptor1.setLevel(HttpLoggingInterceptor.Level.BODY);
        final OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(requestInterceptor)
                .addInterceptor(loggingInterceptor1)
                .build();

        final Retrofit restAdapter = new Retrofit.Builder()
                .baseUrl(MARVEL_ENDPOINT)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();

        return restAdapter.create(MarvelFetchService.class);
    }

    private static class MarvelTypeAdapterFactory implements TypeAdapterFactory {
        @Override
        public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
            final TypeAdapter<T> delegate = gson.getDelegateAdapter(this, type);
            final TypeAdapter<JsonElement> elementAdapter = gson.getAdapter(JsonElement.class);

            return new TypeAdapter<T>() {
                @Override
                public void write(JsonWriter out, T value) throws IOException {
                    delegate.write(out, value);
                }

                @Override
                public T read(JsonReader in) throws IOException {
                    JsonElement jsonElement = elementAdapter.read(in);
                    if (jsonElement.isJsonObject()) {
                        JsonObject jsonObject = jsonElement.getAsJsonObject();
                        if (jsonObject.has("data")) {
                            JsonObject jsonData = jsonObject.getAsJsonObject("data");
                            if (jsonData.has("results")) {
                                jsonElement = jsonData.get("results");
                            }
                        }
                    }

                    return delegate.fromJsonTree(jsonElement);
                }
            }.nullSafe();
        }
    }

    public static String getHash(final long curTime) {
        final String toHash = curTime + BuildConfig.MARVEL_SECRET_API_KEY + BuildConfig.MARVEL_PUB_API_KEY;
        String hashString;
        try {
            MessageDigest digest = MessageDigest.getInstance("MD5");
            digest.update(toHash.getBytes());
            byte messageDigest[] = digest.digest();

            // Create Hex String
            StringBuilder hexString = new StringBuilder();
            for (byte aMessageDigest : messageDigest) {
                String h = Integer.toHexString(0xFF & aMessageDigest);
                while (h.length() < 2)
                    h = "0" + h;
                hexString.append(h);
            }
            System.out.println(hexString.toString());
            hashString = hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            Log.e(LOG_TAG, "No such algorithm exception");
            return "";
        }
        return hashString;
    }

}

File: app/src/main/java/com/fitaleks/heroesthesaurus/network/NetworkHelper.kt
package com.fitaleks.heroesthesaurus.network

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.lang.reflect.Type
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by alexander on 21.08.15.
 */
object NetworkHelper {
    private val LOG_TAG = NetworkHelper::class.java.simpleName
    private val COMICVINE_ENDPOINT = "http://api.comicvine.com"
    private val MTG_ENDPOINT = "https://api.magicthegathering.io/v1/"

    private val TAG_RESULTS = "results"
    private val TAG_CARDS = "cards"
    private val TAG_SETS = "sets"

    val restAdapter: MtgFetchService
        get() {
            val gson = GsonBuilder()
                    .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
                    .registerTypeAdapterFactory(MTGTypeAdapterFactory())
                    .setDateFormat("yyyy-MM-dd")
                    .registerTypeAdapter(Date::class.java, DateDeserializer())
                    .create()

            val requestInterceptor = Interceptor { chain ->
                var request = chain.request()
                val httpUrl = request.url().newBuilder()
                        .build()
                request = request.newBuilder().url(httpUrl).build()
                chain.proceed(request)
            }

            val loggingInterceptor = HttpLoggingInterceptor()
            loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
            val okHttpClient = OkHttpClient.Builder()
                    .addInterceptor(requestInterceptor)
                    .addInterceptor(loggingInterceptor)
                    .connectTimeout(15, TimeUnit.SECONDS)
                    .readTimeout(15, TimeUnit.SECONDS)
                    .build()

            val restAdapter = Retrofit.Builder()
                    .baseUrl(MTG_ENDPOINT)
                    .client(okHttpClient)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build()

            return restAdapter.create(MtgFetchService::class.java)
        }

    private class DateDeserializer : JsonDeserializer<Date> {
        override fun deserialize(json: JsonElement, typeOfT: Type?, context: JsonDeserializationContext?): Date {
            val dateString = json.asString

            val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
            return simpleDateFormat.parse(dateString)
        }
    }

    private class MTGTypeAdapterFactory : TypeAdapterFactory {
        override fun <T : Any?> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T> {
            val delegate = gson.getDelegateAdapter(this, type)
            val elementAdapter = gson.getAdapter(JsonElement::class.java)
            return object : TypeAdapter<T>() {
                @Throws(IOException::class)
                override fun write(out: JsonWriter, value: T) {
                    delegate.write(out, value)
                }

                @Throws(IOException::class)
                override fun read(`in`: JsonReader): T {
                    var jsonElement = elementAdapter.read(`in`)
                    if (jsonElement.isJsonObject) {
                        val jsonObject = jsonElement.asJsonObject
                        if (jsonObject.has(TAG_CARDS)) {
                            jsonElement = jsonObject.get(TAG_CARDS)
                        } else if (jsonObject.has(TAG_SETS)) {
                            jsonElement = jsonObject.get(TAG_SETS)
                        }
                    }

                    return delegate.fromJsonTree(jsonElement)
                }
            }.nullSafe()
        }
    }

}
